tpe/lib/src/Model.cc

```diff
@@ -85,35 +87,42 @@ Entity &Model::AddModel()
   return *it->second.get();
 }
 
+//////////////////////////////////////////////////
+void Model::SetCanonicalLink(const std::size_t linkId)
+{
+  if (linkId != kNullEntityId)
+  {
+    this->dataPtr->canonicalLinkId = linkId;
+  }
+  else if (this->dataPtr->firstLinkId != kNullEntityId)
+  {
+    this->dataPtr->canonicalLinkId = this->dataPtr->firstLinkId;
+  }
+}
+
 //////////////////////////////////////////////////
 Entity &Model::GetCanonicalLink()
 {
   // return canonical link but make sure it exists
   // todo(anyone) Prevent removal of canonical link in a model?
   Entity &linkEnt = this->GetChildById(this->dataPtr->canonicalLinkId);
   if (linkEnt.GetId() != kNullEntityId)
+  {
     return linkEnt;
+  }
 
   // todo(anyone) the code below does not guarantee that the link returned is
   // the first link defined in SDF since GetChildren returns a std::map, and
   // likewise the use of std::set, do not preserve order.
   std::set<Model *> models;
   for (auto &it : this->GetChildren())
   {
-    // return the first link found as canonical link
-    if (dynamic_cast<Link *>(it.second.get()))
-    {
-      return *it.second;
-    }
     // if child is nested model, store it first and only return nested
     // links if there are no links in this model
-    else
+    Model *model = dynamic_cast<Model *>(it.second.get());
+    if (model)
     {
-      Model *model = dynamic_cast<Model *>(it.second.get());
-      if (model)
-      {
-        models.insert(model);
-      }
+      models.insert(model);
```

Comment:
I think we shouldn't have to insert into a set any more. The logic was previously added because we want to process all the links first before moving onto models but we're no longer checking for links here in this for loop.